### ISSUE TITLE
RFC: retire `torch.testing.assert_allclose` in favor of `torch.testing.assert_close`

### ISSUE THREAD

**Issue by pmeier** [pytorch/pytorch#61844]
## TL;DR

We are planning to deprecate and later remove the undocumented `torch.testing.assert_allclose` in favor of [`torch.testing.assert_close`](https://pytorch.org/docs/stable/testing.html) that was introduced with `torch==1.9.0`. They both have the same purpose with the latter being more strict by default but also more configurable.

## Status quo

`assert_close` exists in the `torch.testing` namespace since [`torch==0.4.0`](https://github.com/pytorch/pytorch/blob/3749c581b79cba49f511b19fa02c0f50fa05b250/torch/testing/__init__.py#L16). It was introduced as testing utility for numerics similar to [`numpy.testing.assert_allclose`](https://numpy.org/doc/stable/reference/generated/numpy.testing.assert_allclose.html), but was never documented.

Although in the meantime `TestCase.assertEqual` took over in most places internally, `torch.testing.assert_allclose` is still used in some places. More importantly though, due to the lack of public tensor testing functions, `torch.testing.assert_allclose` is used in downstream projects. A simple GitHub search turns up approximately [1k hits](https://github.com/search?q=torch.testing.assert_allclose&type=code).

In `torch==1.9.0` we introduced [`torch.testing.assert_close`](https://pytorch.org/docs/stable/testing.html) that has the same purpose but is fully documented. There are a few differences compared to `torch.testing.assert_allclose` that will be showcased next.

## Differences

The design goal of `torch.testing.assert_close` is to make it very clear what is being tested without making the user jump to hoops to achieve this. In general that means `assert_close` is stricter by default than `assert_allclose`, but is highly configurable.

### Default tolerances

`assert_allclose` uses https://github.com/pytorch/pytorch/blob/0263865bfe672eef0c71288b9ec886978a2f208a/torch/testing/_core.py#L411-L415

whereas `assert_close` uses

https://github.com/pytorch/pytorch/blob/0263865bfe672eef0c71288b9ec886978a2f208a/torch/testing/_asserts.py#L30-L38

For both the tolerances are `0` in case the dtype is not in the mapping. That means that `assert_allclose` does **not** check for closeness but rather for equality in case the inputs are complex or of dtype `bfloat16`.

### Input types

`assert_allclose` converts everything to a tensor

https://github.com/pytorch/pytorch/blob/0263865bfe672eef0c71288b9ec886978a2f208a/torch/testing/_core.py#L254-L257

whereas `assert_close` requires a direct relation of the input types:

https://github.com/pytorch/pytorch/blob/0263865bfe672eef0c71288b9ec886978a2f208a/torch/testing/_asserts.py#L573-L581

That means `assert_close` does **not** support comparing a `torch.Tensor` to a `numpy.ndarray` or a sequence of scalars, which is possible with `assert_allclose`. On the other hand this enables `assert_close` to effectively check (nested) containers, i.e. sequences or mappings, and provide a traceback where the failure happened.

### Attribute checking / `equal_nan`

`assert_allclose` does not supported any attribute checking besides the shape, which is necessary anyway to compare the values. `assert_close` supports for example checking the devices, dtypes, or strides for equality (the former two are checked by default). Together with `equal_nan=True`, `assert_allclose` is in general more permissive than `assert_close`.

### Non-default tensors

In contrast to `assert_allclose`, `assert_close` has support for sparse COO / CSR and quantized tensors. These changes are already supported in the nightly releases and will make their way into the next release.

## Upgrade guide

We encourage all users to try `torch.testing.assert_close` as a 1-to-1 replacement for `torch.testing.assert_allclose` if the use case allows it. Otherwise a thin wrapper around `torch.testing.assert_close` can be used to get the exact behavior of `assert_allclose` back:

```python
_DTYPE_PRECISIONS = {
    torch.float16: (1e-3, 1e-3),
    torch.float32: (1e-4, 1e-5),
    torch.float64: (1e-5, 1e-8),
}


def _get_default_rtol_and_atol(actual: torch.Tensor, expected: torch.Tensor) -> Tuple[float, float]:
    actual_rtol, actual_atol = _DTYPE_PRECISIONS.get(actual.dtype, (0.0, 0.0))
    expected_rtol, expected_atol = _DTYPE_PRECISIONS.get(expected.dtype, (0.0, 0.0))
    return max(actual_rtol, expected_rtol), max(actual_atol, expected_atol)


def assert_allclose(
    actual: Any,
    expected: Any,
    rtol: Optional[float] = None,
    atol: Optional[float] = None,
    equal_nan: bool = True,
    msg: str = "",
) -> None:
    if not isinstance(actual, torch.Tensor):
        actual = torch.tensor(actual)
    if not isinstance(expected, torch.Tensor):
        expected = torch.tensor(expected, dtype=actual.dtype)

    if rtol is None and atol is None:
        rtol, atol = _get_default_rtol_and_atol(actual, expected)

    torch.testing.assert_close(
        actual,
        expected,
        rtol=rtol,
        atol=atol,
        equal_nan=equal_nan,
        check_device=True,
        check_dtype=False,
        check_stride=False,
        check_is_coalesced=False,
        msg=msg or None,
    )
```

## RFC

Although `assert_allclose` is not documented at all, the current plan is too deprecate it anyway due its widespread usage (see #61841). Please weigh in here if you disagree with this plan.

cc'ing maintainers of projects in the official PyTorch ecosystem that currently use `torch.testing.assert_allclose` and thus would be affected by its retirement (apologies in advance if I picked the "wrong" maintainer).

- [PyTorch Points 3D](https://github.com/nicolas-chaulet/torch-points3d): @nicolas-chaulet 
- [TorchDrift](https://github.com/torchdrift/torchdrift/): @t-vi
- [AllenNLP](https://github.com/allenai/allennlp): @epwalsh
- [Opacus](https://github.com/pytorch/opacus): @ffuuugor

cc @mruberry @VitalyFedyunin @walterddr

**Comment by t-vi** [comment]
Thank you, @pmeier , for the heads up. I used to adapt my code to the new features, but it seems like the main PyTorch innovations affecting me these days are all new deprecations.

From the ecosystem perspective, to me this means that I'm uninterested in supporting older versions of PyTorch, once I change this to avoid the inevitable deprecation warnings, I'll drop support for the versions that don't have the new blessed way - if I would adapt to deprecations in PyTorch in my code, I'd just get my code all sticky so you can yours clean, and I'm not interested in playing this game.

Again, it's perfectly OK here, and I can see the point, but I'm genuinely concerned by the number and scope of deprecations.

**Comment by epwalsh** [comment]
Thanks for pinging us @pmeier. We only use this internally and it looks like it will be pretty easy to switch over to `assert_close`, so not a problem for us.

**Comment by mruberry** [comment]
> Thank you, @pmeier , for the heads up. I used to adapt my code to the new features, but it seems like the main PyTorch innovations affecting me these days are all new deprecations.

At least this deprecation won't happen for awhile. We don't typically deprecation ops until we have stable alternatives, and assert_close was in prototype in PyTorch 1.9, so the earliest we'd see this deprecation is likely PyTorch 1.11.

**Comment by ffuuugor** [comment]
Thanks for the heads-up, shouldn't be a problem for opacus either

**Comment by ffuuugor** [comment]
Quick clarification: what's the deprecation timeline?

At the moment we have some dev dependencies, which doesn't support PyTorch 1.9 (namely [torchcsprng](https://github.com/pytorch/csprng)). As the recommended replacement `torch.testing.assert_close` is introduced in 1.9, we can't migrate until our dependencies are updated to support 1.9

**Comment by pmeier** [comment]
@ffuuugor We won't deprecate until we consider `torch.testing.assert_close` stable. It currently is considered in a prototype state. So you are looking at atleast 2 more releases until deprecation. Since we deprecate for at least one release cycle, this won't be removed before 1.12.

**Comment by malfet** [comment]
Hmm, I have to say it does not say that API is deprecated in the documentation, does it: https://pytorch.org/docs/stable/testing.html#torch.testing.assert_close

**Comment by pmeier** [comment]
@malfet Not sure I understand. As explained above, `torch.testing.assert_allclose` was never an "official" function precisely because it has never been documented and technically the BC rules don't really apply. Still, it was used extensively because there was no alternative. Due to these special circumstances we decided to go through the deprecation period anyway. 

As of `1.9`, `torch.testing.assert_close` is the official and fully documented replacement.

**Comment by malfet** [comment]
@pmeier imo every function, whose name does not start with `_` should be documented and marked as deprecated. 
But I see this warning was printed whenever one tries to use it:
```
FutureWarning: torch.testing.assert_allclose() is deprecated since 1.12 and will be removed in 1.14
```
which seems totally reasonable, but again, it feels a bit strange to omit the function from the official documentation in 1.10 and 1.11 releases

**Comment by pmeier** [comment]
There seems to bit of confusion here. There are two functions:

1. `torch.testing.assert_allclose`
2. `torch.testing.assert_close`

_I'm going to use "tuple notation" with the index from the list above and the function name to minimize the possibility to mix them up accidentally._

As explained in my top comment, (1, `assert_allclose`) is an old function that has never been documented since its inception in `torch==0.4.0`. Although it saw some usage internally, it was mostly used by third-party projects since they had no access to our `TestCase.assertEqual` but still needed a way to compare floating point tensors.

When we started to work on `torch.testing` ~1.5 years ago, we realized that unofficial and undocumented function is used everywhere. Since it had some undesirable features as explained in my top comment, we settled on adding a new function (2, `assert_close`) that should ultimately replace (1, `assert_allclose`). 

This is what this RFC is for. The acceptance of it by the community is what lead to the deprecation of (1, `assert_allclose`) in favor of (2, `assert_close`) in #73348 and ultimately the removal after the deprecation period in #87974.

So (1, `assert_allclose`) was never documented in the first place since it was added ~5 years ago. In contrast, (2, `assert_close`) was documented from the beginning, i.e. `torch==1.9.0` and has never been omitted or removed from the documentation.

**Comment by bluenote10** [comment]
What is a bit unfortunate is that even in pytorch 1.13.1, pyright (i.e., VSCode) complains about usages of `torch.testing.assert_close` but happily offers the deprecated `torch.testing.assert_allclose` via its auto-completion:

![image](https://user-images.githubusercontent.com/3620703/220886426-6af8df8f-b152-4858-a85d-5ff7ce95d54f.png)

The error says `"assert_close" is not exported from module "torch.testing"PylancereportPrivateImportUsage`.

This could indicate that something isn't quite right about the re-export of `assert_close` yet?

**Comment by pmeier** [comment]
I've never used pyright, so not sure what it is complaining about here, but we aren't doing anything special here:

https://github.com/pytorch/pytorch/blob/49444c3e546bf240bed24a101e747422d1f8a0ee/torch/testing/__init__.py#L1-L4

**Comment by bluenote10** [comment]
I think the problem is that pyright / pylance / VSCode strictly sticks to the [specs here](https://github.com/python/typing/blob/master/docs/source/libraries.rst#library-interface-public-and-private-symbols), in particular the second bullet point (_imported symbols are considered private by default, unless ..._). The specs seem to be a bit under-specified regarding the `from ._module` case, i.e., whether the fact that a module is "marked" as private should be respected. In any case, I did a quick test: The problem can be fixed by changing the first line of the `__init__.py` to use a redundant `as` to signal non-privateness of the symbol, i.e.:

```py
from ._comparison import assert_close as assert_close
```

Now, auto-completion works, and VSCode no longer shows an error. Alternatively explicit `__all__` specifying could be used, but that's tedious.

To avoid exposing the "deprecated" symbols publicly, one could turn the wildcard import in line 4 into an explicit import of the symbols, which would make them private, i.e., they could be used at runtime, but VSCode users would not see them in the auto-completion.

(related upstream issue: https://github.com/microsoft/pylance-release/issues/2200)

**Comment by pmeier** [comment]
That is already done for the next release, which will come out in a few weeks:

https://github.com/pytorch/pytorch/blob/3b966a6ce3d39122998a362c2b4cb95e34a79d0b/torch/testing/__init__.py#L1-L3
